fast64_internal/f3d/flipbook.py

```diff
@@ -0,0 +1,329 @@
+                        setTexNodeImage(material, i, flipbookIndex)
+
+
+# END GAME SPECIFIC CALLBACKS
+
+# we use a handler since update functions are not called when a property is animated.
+@persistent
+def flipbookAnimHandler(dummy):
+    if bpy.context.scene.gameEditorMode == "OOT":
+        for obj in bpy.data.objects:
+            if isinstance(obj.data, bpy.types.Armature):
+                # we only want to update texture on keyframed armatures.
+                # this somewhat mitigates the issue of two skeletons using the same flipbook material.
+                if obj.animation_data is None or obj.animation_data.action is None:
+                    return
+                action = obj.animation_data.action
+                if not (
+                    action.fcurves.find("ootLinkTextureAnim.eyes") is None
+                    or action.fcurves.find("ootLinkTextureAnim.mouth") is None
+                ):
+                    ootFlipbookAnimUpdate(obj.data, obj, "8", obj.ootLinkTextureAnim.eyes)
+                    ootFlipbookAnimUpdate(obj.data, obj, "9", obj.ootLinkTextureAnim.mouth)
+    else:
+        pass
+
+
+class Flipbook_MaterialPanel(bpy.types.Panel):
+    bl_label = "Flipbook Material"
+    bl_idname = "MATERIAL_PT_Flipbook_Material_Inspector"
+    bl_space_type = "PROPERTIES"
+    bl_region_type = "WINDOW"
+    bl_context = "material"
+    bl_options = {"HIDE_HEADER"}
+
+    @classmethod
+    def poll(cls, context):
+        return context.material is not None and context.scene.gameEditorMode in ["OOT"]
+
+    def draw(self, context):
+        layout = self.layout
+        mat = context.material
+        col = layout.column()
+
+        if context.scene.gameEditorMode == "OOT":
+            checkFlipbookReference = ootFlipbookReferenceIsValid
+            flipbookRequirementMessage = ootFlipbookRequirementMessage
+        else:
+            checkFlipbookReference = None
+            flipbookRequirementMessage = None
+
+        drawFlipbookGroupProperty(col.box().column(), mat, checkFlipbookReference, flipbookRequirementMessage)
+
+
+def setTexNodeImage(material: bpy.types.Material, texIndex: int, flipbookIndex: int):
+    flipbook = getattr(material.flipbookGroup, "flipbook" + str(texIndex))
+    for texNode in iter_tex_nodes(material.node_tree, texIndex):
+        if texNode.image is not flipbook.textures[flipbookIndex].image:
+            texNode.image = flipbook.textures[flipbookIndex].image
+
+
+flipbook_classes = [
+    ImagePointerProperty,
+    AddFlipbookTexture,
+    RemoveFlipbookTexture,
+    MoveFlipbookTexture,
+    VisualizeFlipbookTexture,
+    FlipbookProperty,
+    FlipbookGroupProperty,
+    Flipbook_MaterialPanel,
+]
+
+
+def flipbook_register():
+    for cls in flipbook_classes:
+        register_class(cls)
+
+    bpy.app.handlers.frame_change_pre.append(flipbookAnimHandler)
+    bpy.types.Material.flipbookGroup = bpy.props.PointerProperty(type=FlipbookGroupProperty)
```

Comment:
I was thinking that too, but its not really related to F3D at all so I think its better to have it separate.

Reply:
we could have a `material.fast64` `Fast64_MaterialProperties` though, maybe?

Like
- `scene.fast64` `Fast64_Properties`
- `bone.fast64` `Fast64_BoneProperties`
- `object.fast64` `Fast64_ObjectProperties`

Reply:
Honestly I'm kinda iffy on the bone/object versions in general. Most of those properties are already outside the fast64 property groups, and can't be changed easily since people are already using them (unlike scene, which is mostly operator settings). If some properties are in the fast64 property group and some aren't, then it defeats the purpose of having a fast64 property group in the first place.